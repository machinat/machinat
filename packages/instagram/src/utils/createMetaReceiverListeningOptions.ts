import { PopEventWrapper } from '@sociably/core';
import { ListeningPlatformOptions } from '@sociably/meta-api';
import InstagramBot from '../Bot.js';
import { INSTAGRAM } from '../constant.js';
import eventFactory from '../event/factory.js';
import { InstagramEventContext } from '../types.js';

const createMetaReceiverListeningOptions = (
  bot: InstagramBot,
  popEventWrapper: PopEventWrapper<InstagramEventContext, null>
): ListeningPlatformOptions<InstagramEventContext> => ({
  platform: INSTAGRAM,
  bot,
  objectType: 'instagram',
  makeEventsFromUpdate: (updateData) => {
    const { id: pageId, messaging, stanby } = updateData;
    const isStandby = stanby !== undefined;
    const rawEvents = isStandby ? stanby : messaging;

    return rawEvents.map((rawEvent) =>
      eventFactory(pageId, isStandby, rawEvent)
    );
  },
  popEvent: popEventWrapper(async () => null),
});

export default createMetaReceiverListeningOptions;
