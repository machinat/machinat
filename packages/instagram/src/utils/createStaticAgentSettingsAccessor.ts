import { InstagramAgentSettings } from '../types.js';
import { AgentSettingsAccessorI } from '../interface.js';

const createStaticAgentSettingsAccessor = (
  settings: InstagramAgentSettings[]
): AgentSettingsAccessorI => ({
  getAgentSettings: async (page) =>
    settings.find(({ pageId }) => pageId === page.id) || null,
  getAgentSettingsBatch: async (pages) =>
    pages.map(
      (page) => settings.find(({ pageId }) => pageId === page.id) || null
    ),
});

export default createStaticAgentSettingsAccessor;
