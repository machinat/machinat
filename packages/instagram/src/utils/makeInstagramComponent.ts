import { makeNativeComponent } from '@sociably/core/renderer';
import { INSTAGRAM } from '../constant.js';

export default makeNativeComponent(INSTAGRAM);
