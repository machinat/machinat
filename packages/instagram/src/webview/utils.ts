/* eslint-disable import/prefer-default-export */
import type { ContextDetails } from '@sociably/auth';
import InstagramChat from '../Chat.js';
import InstagramUser from '../User.js';
import InstagramAgent from '../Agent.js';
import UserProfile from '../UserProfile.js';
import type { InstagramAuthContext, InstagramAuthData } from './types.js';

export const getAuthContextDetails = (
  data: InstagramAuthData
): ContextDetails<InstagramAuthContext> => {
  const {
    agent: { id: agentId, name: agentUsername },
    user: userId,
    profile: profileData,
  } = data;
  return {
    channel: new InstagramAgent(agentId, agentUsername),
    thread: new InstagramChat(agentId, { id: userId }),
    user: new InstagramUser(agentId, userId),
    userProfile: profileData ? new UserProfile(profileData) : null,
    agentUsername,
  };
};
