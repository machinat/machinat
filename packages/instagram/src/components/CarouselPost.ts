import { SociablyNode } from '@sociably/core';
import { makeUnitSegment, UnitSegment } from '@sociably/core/renderer';
import { formatNode } from '@sociably/core/utils';
import makeInstagramComponent from '../utils/makeInstagramComponent.js';
import { InstagramComponent, PostSegmentValue } from '../types.js';

export type CarouselPostProps = {
  /**
   * A caption for the carousel. Can include hashtags (example:
   * #crazywildebeest) and usernames of Instagram users (example: @natgeo).
   *
   * Mentioned Instagram users receive a notification when the container is
   * published. Maximum 2200 characters, 30 hashtags, and 20 @ tags.
   */
  caption?: string;
  /**
   * An array of up to 10 {@link ImagePost} or {@link VideoPost} of each image and
   * video that should appear in the published carousel. Carousels can have up
   * to 10 total images, vidoes, or a mix of the two.
   */
  children: SociablyNode;
  /**
   * The ID of a Page associated with a location that you want to tag the image
   * or video with.
   *
   * Use the [Pages Search
   * API](https://developers.facebook.com/micro_site/url/?click_from_context_menu=true&country=TW&destination=https%3A%2F%2Fdevelopers.facebook.com%2Fdocs%2Fpages%2Fsearching&event_type=click&last_nav_impression_id=0P14xS35NFdHjODVp&max_percent_page_viewed=100&max_viewport_height_px=859&max_viewport_width_px=1512&orig_http_referrer=https%3A%2F%2Fdevelopers.facebook.com%2Fdocs%2Finstagram-api%2Freference%2Fig-user%2Fmedia&orig_request_uri=https%3A%2F%2Fdevelopers.facebook.com%2Fajax%2Fpagelet%2Fgeneric.php%2FDeveloperNotificationsPayloadPagelet%3Ffb_dtsg_ag%3D--sanitized--%26data%3D%257B%2522businessUserID%2522%253Anull%252C%2522cursor%2522%253Anull%252C%2522length%2522%253A15%252C%2522clientRequestID%2522%253A%2522js_xp%2522%257D%26__usid%3D6-Ts0vsi64krspa%253APs1w52q117t3d8%253A0-As1w323s9fw4p-RV%253D6%253AF%253D%26jazoest%3D24916&region=apac&scrolled=true&session_id=1dfuyoonWzZNO93Ez&site=developers)
   * to search for Pages whose names match a search string, then parse the
   * results to identify any Pages that have been created for a physical
   * location. Include the location field in your query and verify that the Page
   * you want to use has location data. Attempting to create a container using a
   * Page that has no location data will fail with coded exception
   * INSTAGRAM_PLATFORM_API__INVALID_LOCATION_ID.
   *
   * Not supported on images or videos in carousels.
   */
  locationId?: string;
};

/**
 * The log out button triggers the account unlinking flow.
 *
 * @category Component
 * @props {@link MediaProps}
 * @guides Check official send API [doc](https://developers.facebook.com/docs/messenger-platform/send-messages)
 *   and [reference](https://developers.facebook.com/docs/messenger-platform/reference/send-api).
 */
export const CarouselPost: InstagramComponent<
  CarouselPostProps,
  UnitSegment<PostSegmentValue>
> = makeInstagramComponent(async function CarouselPost(node, path, render) {
  const { caption, children, locationId } = node.props;

  const mediaSegments = await render<PostSegmentValue, never>(children, path);
  if (!mediaSegments?.length) {
    throw new Error(
      `<CarouselPost /> must contain at least one <ImagePost/> or <VideoPost/>`,
    );
  }
  for (const segment of mediaSegments) {
    if (
      (segment.type !== 'unit' && segment.type !== 'raw') ||
      segment.value.type !== 'post' ||
      (segment.value.params.media_type !== undefined &&
        segment.value.params.media_type !== 'VIDEO')
    ) {
      throw new Error(
        `${formatNode(
          segment.node,
        )} is not a valid children of <CarouselPost />, must be <ImagePost/> or <VideoPost/>`,
      );
    }
  }

  return [
    makeUnitSegment(node, path, {
      type: 'post',
      params: {
        media_type: 'CAROUSEL',
        caption,
        location_id: locationId,
        children: mediaSegments.map(
          (segment) => (segment.value as PostSegmentValue).params,
        ),
      },
    }),
  ];
});
