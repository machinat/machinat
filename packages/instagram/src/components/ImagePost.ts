import { makeUnitSegment, UnitSegment } from '@sociably/core/renderer';
import makeInstagramComponent from '../utils/makeInstagramComponent.js';
import { InstagramComponent, PostSegmentValue } from '../types.js';

export type ImagePostProps = {
  /** The URL to the image file */
  url: string;
  /**
   * A caption for the image. Can include hashtags (example: #crazywildebeest)
   * and usernames of Instagram users (example: @natgeo).
   *
   * Mentioned Instagram users receive a notification when the container is
   * published. Maximum 2200 characters, 30 hashtags, and 20 @ tags.
   *
   * Not supported on images in carousels.
   */
  caption?: string;
  /**
   * The ID of a Page associated with a location that you want to tag the image
   * with.
   *
   * Use the [Pages Search
   * API](https://developers.facebook.com/micro_site/url/?click_from_context_menu=true&country=TW&destination=https%3A%2F%2Fdevelopers.facebook.com%2Fdocs%2Fpages%2Fsearching&event_type=click&last_nav_impression_id=0P14xS35NFdHjODVp&max_percent_page_viewed=100&max_viewport_height_px=859&max_viewport_width_px=1512&orig_http_referrer=https%3A%2F%2Fdevelopers.facebook.com%2Fdocs%2Finstagram-api%2Freference%2Fig-user%2Fmedia&orig_request_uri=https%3A%2F%2Fdevelopers.facebook.com%2Fajax%2Fpagelet%2Fgeneric.php%2FDeveloperNotificationsPayloadPagelet%3Ffb_dtsg_ag%3D--sanitized--%26data%3D%257B%2522businessUserID%2522%253Anull%252C%2522cursor%2522%253Anull%252C%2522length%2522%253A15%252C%2522clientRequestID%2522%253A%2522js_xp%2522%257D%26__usid%3D6-Ts0vsi64krspa%253APs1w52q117t3d8%253A0-As1w323s9fw4p-RV%253D6%253AF%253D%26jazoest%3D24916&region=apac&scrolled=true&session_id=1dfuyoonWzZNO93Ez&site=developers)
   * to search for Pages whose names match a search string, then parse the
   * results to identify any Pages that have been created for a physical
   * location. Include the location field in your query and verify that the Page
   * you want to use has location data. Attempting to create a container using a
   * Page that has no location data will fail with coded exception
   * INSTAGRAM_PLATFORM_API__INVALID_LOCATION_ID.
   *
   * Not supported on images in carousels.
   */
  locationId?: string;
  /**
   * An array of objects specifying which product tags to tag the image with
   * (maximum of 5; tags and product IDs must be unique). Each object should
   * have the following information
   */
  productTags?: {
    /** The product ID. */
    productId: string;
    /**
     * An optional float that indicates percentage distance from left edge of
     * the published media image. Value must be within 0.0–1.0 range.
     */
    x?: number;
    /**
     * An optional float that indicates percentage distance from top edge of the
     * published media image. Value must be within 0.0–1.0 range.
     */
    y?: number;
  }[];
  /**
   * An array of public usernames and x/y coordinates for any public Instagram
   * users who you want to tag in the image. Each object should have the
   * following information:
   */
  userTags?: {
    /** Public usernames */
    usernames: string[];
    /**
     * An optional float that indicates percentage distance from left edge of
     * the published media image. Value must be within 0.0–1.0 range.
     */
    x?: number;
    /**
     * An optional float that indicates percentage distance from top edge of the
     * published media image. Value must be within 0.0–1.0 range.
     */
    y?: number;
  }[];
};

/**
 * Post a image post to Instagram as the agent.
 *
 * @category Component
 * @props {@link ImagePostProps}
 * @guides Check official API [doc](https://developers.facebook.com/docs/instagram-api/guides/content-publishing)
 *   and [reference](https://developers.facebook.com/docs/instagram-api/reference/ig-user/media).
 */
export const ImagePost: InstagramComponent<
  ImagePostProps,
  UnitSegment<PostSegmentValue>
> = makeInstagramComponent(function ImagePost(node, path) {
  const { url, caption, locationId, productTags, userTags } = node.props;

  return [
    makeUnitSegment(node, path, {
      type: 'post',
      params: {
        image_url: url,
        caption,
        location_id: locationId,
        product_tags: productTags,
        user_tags: userTags,
      },
    }),
  ];
});
