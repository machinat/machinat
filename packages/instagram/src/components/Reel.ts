import { makeUnitSegment, UnitSegment } from '@sociably/core/renderer';
import makeInstagramComponent from '../utils/makeInstagramComponent.js';
import { InstagramComponent, PostSegmentValue } from '../types.js';

export type ReelProps = {
  /** The URL to the video file */
  url: string;
  /**
   * Name of the audio of your Reels media. You can only rename once, either
   * while creating a reel or after from the audio page.
   */
  audioName?: string;
  /**
   * The path to an image to use as the cover image for the Reels tab. We will
   * cURL the image using the URL that you specify so the image must be on a
   * public server. If you specify both coverUrl and thumbOffset, we use
   * coverUrl and ignore thumbOffset. The image must conform to the
   * specifications for a Reels cover photo.
   */
  coverUrl?: string;
  /**
   * The ID of a Page associated with a location that you want to tag the reel
   * with.
   *
   * Use the [Pages Search
   * API](https://developers.facebook.com/micro_site/url/?click_from_context_menu=true&country=TW&destination=https%3A%2F%2Fdevelopers.facebook.com%2Fdocs%2Fpages%2Fsearching&event_type=click&last_nav_impression_id=0P14xS35NFdHjODVp&max_percent_page_viewed=100&max_viewport_height_px=859&max_viewport_width_px=1512&orig_http_referrer=https%3A%2F%2Fdevelopers.facebook.com%2Fdocs%2Finstagram-api%2Freference%2Fig-user%2Fmedia&orig_request_uri=https%3A%2F%2Fdevelopers.facebook.com%2Fajax%2Fpagelet%2Fgeneric.php%2FDeveloperNotificationsPayloadPagelet%3Ffb_dtsg_ag%3D--sanitized--%26data%3D%257B%2522businessUserID%2522%253Anull%252C%2522cursor%2522%253Anull%252C%2522length%2522%253A15%252C%2522clientRequestID%2522%253A%2522js_xp%2522%257D%26__usid%3D6-Ts0vsi64krspa%253APs1w52q117t3d8%253A0-As1w323s9fw4p-RV%253D6%253AF%253D%26jazoest%3D24916&region=apac&scrolled=true&session_id=1dfuyoonWzZNO93Ez&site=developers)
   * to search for Pages whose names match a search string, then parse the
   * results to identify any Pages that have been created for a physical
   * location. Include the location field in your query and verify that the Page
   * you want to use has location data. Attempting to create a container using a
   * Page that has no location data will fail with coded exception
   * INSTAGRAM_PLATFORM_API__INVALID_LOCATION_ID.
   */
  locationId?: string;
  /**
   * When `true`, indicates that the reel can appear in both the Feed and Reels
   * tabs. When `false`, indicates the reel can only appear in the Reels tab.
   */
  shareToFeed?: boolean;
  /**
   * Location, in milliseconds, of the reel video frame to be used as the cover
   * thumbnail image. The default value is 0, which is the first frame of the
   * video or reel. For reels, if you specify both cover_url and thumb_offset,
   * we use cover_url and ignore thumb_offset.
   */
  thumbOffset?: number;
};

/**
 * Post a Reels video to Instagram as the agent.
 *
 * @category Component
 * @props {@link ReelProps}
 * @guides Check official API [doc](https://developers.facebook.com/docs/instagram-api/guides/content-publishing)
 *   and [reference](https://developers.facebook.com/docs/instagram-api/reference/ig-user/media).
 */
export const Reel: InstagramComponent<
  ReelProps,
  UnitSegment<PostSegmentValue>
> = makeInstagramComponent(function Reel(node, path) {
  const { url, audioName, coverUrl, locationId, shareToFeed, thumbOffset } =
    node.props;

  return [
    makeUnitSegment(node, path, {
      type: 'post',
      params: {
        media_type: 'REELS',
        video_url: url,
        audio_name: audioName,
        cover_url: coverUrl,
        location_id: locationId,
        share_to_feed: shareToFeed,
        thumb_offset: thumbOffset,
      },
    }),
  ];
});
