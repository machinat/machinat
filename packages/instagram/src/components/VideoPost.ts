import { makeUnitSegment, UnitSegment } from '@sociably/core/renderer';
import makeInstagramComponent from '../utils/makeInstagramComponent.js';
import { InstagramComponent, PostSegmentValue } from '../types.js';

export type VideoPostProps = {
  /** The URL to the video file */
  url: string;
  /**
   * A caption for the video. Can include hashtags (example: #crazywildebeest)
   * and usernames of Instagram users (example: @natgeo).
   *
   * Mentioned Instagram users receive a notification when the container is
   * published. Maximum 2200 characters, 30 hashtags, and 20 @ tags.
   *
   * Not supported on videos in carousels.
   */
  caption?: string;
  /**
   * The ID of a Page associated with a location that you want to tag the video
   * with.
   *
   * Use the [Pages Search
   * API](https://developers.facebook.com/micro_site/url/?click_from_context_menu=true&country=TW&destination=https%3A%2F%2Fdevelopers.facebook.com%2Fdocs%2Fpages%2Fsearching&event_type=click&last_nav_impression_id=0P14xS35NFdHjODVp&max_percent_page_viewed=100&max_viewport_height_px=859&max_viewport_width_px=1512&orig_http_referrer=https%3A%2F%2Fdevelopers.facebook.com%2Fdocs%2Finstagram-api%2Freference%2Fig-user%2Fmedia&orig_request_uri=https%3A%2F%2Fdevelopers.facebook.com%2Fajax%2Fpagelet%2Fgeneric.php%2FDeveloperNotificationsPayloadPagelet%3Ffb_dtsg_ag%3D--sanitized--%26data%3D%257B%2522businessUserID%2522%253Anull%252C%2522cursor%2522%253Anull%252C%2522length%2522%253A15%252C%2522clientRequestID%2522%253A%2522js_xp%2522%257D%26__usid%3D6-Ts0vsi64krspa%253APs1w52q117t3d8%253A0-As1w323s9fw4p-RV%253D6%253AF%253D%26jazoest%3D24916&region=apac&scrolled=true&session_id=1dfuyoonWzZNO93Ez&site=developers)
   * to search for Pages whose names match a search string, then parse the
   * results to identify any Pages that have been created for a physical
   * location. Include the location field in your query and verify that the Page
   * you want to use has location data. Attempting to create a container using a
   * Page that has no location data will fail with coded exception
   * INSTAGRAM_PLATFORM_API__INVALID_LOCATION_ID.
   *
   * Not supported on videos in carousels.
   */
  locationId?: string;
  /**
   * An array of objects specifying which product tags to tag the video with
   * (maximum of 5; tags and product IDs must be unique). Each object should
   * have the following information
   */
  productTags?: {
    /** The product ID. */
    productId: string;
  }[];
  /**
   * Location, in milliseconds, of the video frame to be used as the cover
   * thumbnail image. The default value is 0, which is the first frame of the
   * video or reel. For reels, if you specify both cover_url and thumb_offset,
   * we use cover_url and ignore thumb_offset.
   */
  thumbOffset?: number;
  /**
   * An array of public usernames and x/y coordinates for any public Instagram
   * users who you want to tag in the video. Each object should have the
   * following information:
   */
  userTags?: {
    /** Public usernames */
    usernames: string[];
  }[];
};

/**
 * Post a video post to Instagram as the agent.
 *
 * @category Component
 * @props {@link VideoPostProps}
 * @guides Check official API [doc](https://developers.facebook.com/docs/instagram-api/guides/content-publishing)
 *   and [reference](https://developers.facebook.com/docs/instagram-api/reference/ig-user/media).
 */
export const VideoPost: InstagramComponent<
  VideoPostProps,
  UnitSegment<PostSegmentValue>
> = makeInstagramComponent(function VideoPost(node, path) {
  const { url, caption, locationId, productTags, thumbOffset, userTags } =
    node.props;

  return [
    makeUnitSegment(node, path, {
      type: 'post',
      params: {
        media_type: 'VIDEO',
        video_url: url,
        caption,
        location_id: locationId,
        product_tags: productTags,
        thumb_offset: thumbOffset,
        user_tags: userTags,
      },
    }),
  ];
});
