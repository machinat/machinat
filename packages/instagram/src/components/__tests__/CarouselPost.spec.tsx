import Sociably from '@sociably/core';
import { isNativeType } from '@sociably/core/utils';
import { ImagePost } from '../ImagePost.js';
import { VideoPost } from '../VideoPost.js';
import { Reel } from '../Reel.js';
import { CarouselPost } from '../CarouselPost.js';
import { TypingOn } from '../index.js';
import { renderUnitElement } from './utils.js';

it('is valid root Component', () => {
  expect(isNativeType(<CarouselPost> </CarouselPost>)).toBe(true);
  expect(CarouselPost.$$platform).toBe('instagram');
  expect(CarouselPost.$$name).toBe('CarouselPost');
});

it('match snapshot', async () => {
  await expect(
    renderUnitElement(
      <CarouselPost>
        <ImagePost url="http://abc.com/..." />
        <ImagePost url="http://xyz.com/..." />
      </CarouselPost>,
    ),
  ).resolves.toMatchSnapshot();

  await expect(
    renderUnitElement(
      <CarouselPost
        caption="This is a #carousel post with #videos"
        locationId="1234567890"
      >
        <VideoPost url="http://abc.com/..." />
        <VideoPost url="http://xyz.com/..." />
      </CarouselPost>,
    ),
  ).resolves.toMatchSnapshot();
});

it('throw if there is no items in children', async () => {
  await expect(
    renderUnitElement(<CarouselPost>{null}</CarouselPost>),
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"<CarouselPost /> must contain at least one <ImagePost/> or <VideoPost/>"`,
  );
  await expect(
    renderUnitElement(<CarouselPost>{[]}</CarouselPost>),
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"<CarouselPost /> must contain at least one <ImagePost/> or <VideoPost/>"`,
  );
});

it('throw if "children" prop contains invalid items', async () => {
  await expect(
    renderUnitElement(<CarouselPost>foo</CarouselPost>),
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `""foo" is not a valid children of <CarouselPost />, must be <ImagePost/> or <VideoPost/>"`,
  );
  await expect(
    renderUnitElement(
      <CarouselPost>
        <Sociably.Pause />
      </CarouselPost>,
    ),
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"<Pause /> is not a valid children of <CarouselPost />, must be <ImagePost/> or <VideoPost/>"`,
  );
  await expect(
    renderUnitElement(
      <CarouselPost>
        <Reel url="http://..." />
      </CarouselPost>,
    ),
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"<Reel /> is not a valid children of <CarouselPost />, must be <ImagePost/> or <VideoPost/>"`,
  );
  await expect(
    renderUnitElement(
      <CarouselPost>
        <TypingOn />
      </CarouselPost>,
    ),
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"<TypingOn /> is not a valid children of <CarouselPost />, must be <ImagePost/> or <VideoPost/>"`,
  );
});
