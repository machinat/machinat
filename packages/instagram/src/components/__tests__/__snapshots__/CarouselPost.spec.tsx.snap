// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match snapshot 1`] = `
[
  {
    "node": <CarouselPost>
      <ImagePost
        url="http://abc.com/..."
      />
      <ImagePost
        url="http://xyz.com/..."
      />
    </CarouselPost>,
    "path": "$",
    "type": "unit",
    "value": {
      "params": {
        "caption": undefined,
        "children": [
          {
            "caption": undefined,
            "image_url": "http://abc.com/...",
            "location_id": undefined,
            "product_tags": undefined,
            "user_tags": undefined,
          },
          {
            "caption": undefined,
            "image_url": "http://xyz.com/...",
            "location_id": undefined,
            "product_tags": undefined,
            "user_tags": undefined,
          },
        ],
        "location_id": undefined,
        "media_type": "CAROUSEL",
      },
      "type": "post",
    },
  },
]
`;

exports[`match snapshot 2`] = `
[
  {
    "node": <CarouselPost
      caption="This is a #carousel post with #videos"
      locationId="1234567890"
    >
      <VideoPost
        url="http://abc.com/..."
      />
      <VideoPost
        url="http://xyz.com/..."
      />
    </CarouselPost>,
    "path": "$",
    "type": "unit",
    "value": {
      "params": {
        "caption": "This is a #carousel post with #videos",
        "children": [
          {
            "caption": undefined,
            "location_id": undefined,
            "media_type": "VIDEO",
            "product_tags": undefined,
            "thumb_offset": undefined,
            "user_tags": undefined,
            "video_url": "http://abc.com/...",
          },
          {
            "caption": undefined,
            "location_id": undefined,
            "media_type": "VIDEO",
            "product_tags": undefined,
            "thumb_offset": undefined,
            "user_tags": undefined,
            "video_url": "http://xyz.com/...",
          },
        ],
        "location_id": "1234567890",
        "media_type": "CAROUSEL",
      },
      "type": "post",
    },
  },
]
`;
